A file with a TestMain() and no Test() results in a compilation error.

I had a file that looked like this: 

```go
package e2e_kubeadm

import (
	"flag"
	"os"
	"testing"

	"github.com/spf13/pflag"
	"k8s.io/kubernetes/test/e2e/framework"
)

func init() {
	framework.RegisterCommonFlags()

	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
}

func TestMain(m *testing.M) {
	pflag.Parse()
	framework.AfterReadingAllFlags(&framework.TestContext)
	os.Exit(m.Run())
}
```

But if failed to compile: 
```
/private/var/tmp/_bazel_liz/aaf126f19606a5dec6111ea1f19477ba/bazel-sandbox/3918368524966085086/execroot/__main__/bazel-out/darwin-fastbuild/bin/test/e2e_kubeadm/darwin_amd64_stripped/go_default_test~/testmain.go:109:2: undefined: l
```

That file looks like 
```go
package main
import (
	"flag"
	"log"
	"os"
	"fmt"
	"strconv"
	"testing"
	"testing/internal/testdeps"




)

var allTests = []testing.InternalTest{

}

var benchmarks = []testing.InternalBenchmark{

}

var examples = []testing.InternalExample{

}

func testsInShard() []testing.InternalTest {
	totalShards, err := strconv.Atoi(os.Getenv("TEST_TOTAL_SHARDS"))
	if err != nil || totalShards <= 1 {
		return allTests
	}
	shardIndex, err := strconv.Atoi(os.Getenv("TEST_SHARD_INDEX"))
	if err != nil || shardIndex < 0 {
		return allTests
	}
	tests := []testing.InternalTest{}
	for i, t := range allTests {
		if i % totalShards == shardIndex {
			tests = append(tests, t)
		}
	}
	return tests
}

func coverRegisterAll() testing.Cover {
	coverage := testing.Cover{
		Mode: "set",
		CoveredPackages: "",
		Counters: map[string][]uint32{},
		Blocks: map[string][]testing.CoverBlock{},
	}

	return coverage
}

func coverRegisterFile(coverage *testing.Cover, fileName string, counter []uint32, pos []uint32, numStmts []uint16) {
	if 3*len(counter) != len(pos) || len(counter) != len(numStmts) {
		panic("coverage: mismatched sizes")
	}
	if coverage.Counters[fileName] != nil {
		// Already registered.
		fmt.Printf("Already covered %s\n", fileName)
		return
	}
	coverage.Counters[fileName] = counter
	block := make([]testing.CoverBlock, len(counter))
	for i := range counter {
		block[i] = testing.CoverBlock{
			Line0: pos[3*i+0],
			Col0: uint16(pos[3*i+2]),
			Line1: pos[3*i+1],
			Col1: uint16(pos[3*i+2]>>16),
			Stmts: numStmts[i],
		}
	}
	coverage.Blocks[fileName] = block
}

func main() {
	// Check if we're being run by Bazel and change directories if so.
	// TEST_SRCDIR is set by the Bazel test runner, so that makes a decent proxy.
	if _, ok := os.LookupEnv("TEST_SRCDIR"); ok {
		if err := os.Chdir("test/e2e_kubeadm"); err != nil {
			log.Fatalf("could not change to test directory: %v", err)
		}
	}

	if filter := os.Getenv("TESTBRIDGE_TEST_ONLY"); filter != "" {
		if f := flag.Lookup("test.run"); f != nil {
			f.Value.Set(filter)
		}
	}

	coverage := coverRegisterAll()
	if len(coverage.Counters) > 0 {
		testing.RegisterCover(coverage)
	}

	if coverageDat, ok := os.LookupEnv("COVERAGE_OUTPUT_FILE"); ok {
		if testing.CoverMode() != "" {
			flag.Lookup("test.coverprofile").Value.Set(coverageDat)
		}
	}

	m := testing.MainStart(testdeps.TestDeps{}, testsInShard(), benchmarks, examples)
	
	l.TestMain(m)
	
}
```

The `l` in `l.TestMain` should be imported by generate_test_main, but the file is never included unless it also has a regular Test* case.

--- a/go/tools/builders/generate_test_main.go
+++ b/go/tools/builders/generate_test_main.go
@@ -322,6 +322,7 @@ func run(args []string) error {
 			}
 			if fn.Name.Name == "TestMain" {
 				// TestMain is not, itself, a test
+				pkgs[pkg] = true
 				cases.TestMain = fmt.Sprintf("%s.%s", pkg, fn.Name.Name)
 				continue
 			}

--- a/tests/core/go_test/BUILD.bazel
+++ b/tests/core/go_test/BUILD.bazel
@@ -69,3 +69,9 @@ go_test(
         "Expected": "Example",
     },
 )
+
+go_test(
+    name = "only_testmain_test",
+    size = "small",
+    srcs = ["only_testmain_test.go"],
+)

--- a/tests/core/go_test/README.rst
+++ b/tests/core/go_test/README.rst
@@ -3,14 +3,14 @@ Basic go_test functionality
 
 .. _go_test: /go/core.rst#_go_test
 
-Tests to ensure that basic features of go_test_ are working as expected.
+Tests to ensure that basic features of `go_test`_ are working as expected.
 
 .. contents::
 
 internal_test
 -------------
 
-Test that a go_test_ rule that adds white box tests to an embedded package works.
+Test that a `go_test`_ rule that adds white box tests to an embedded package works.
 This builds a library with `lib.go <lib.go>`_ and then a package with an
 `internal test <internal_test.go>`_ that contains the test case.
 It uses x_def stamped values to verify the library names are correct.
@@ -18,14 +18,14 @@ It uses x_def stamped values to verify the library names are correct.
 external_test
 -------------
 
-Test that a go_test_ rule that adds black box tests for a dependant package works.
+Test that a `go_test`_ rule that adds black box tests for a dependant package works.
 This builds a library with `lib.go <lib.go>`_ and then a package with an
 `external test <external_test.go>`_ that contains the test case.
 It uses x_def stamped values to verify the library names are correct.
 
 combined_test
 -------------
-Test that a go_test_ rule that adds both white and black box tests for a
+Test that a `go_test`_ rule that adds both white and black box tests for a
 package works.
 This builds a library with `lib.go <lib.go>`_ and then a one merged with the
 `internal test <internal_test.go>`_, and then another one that depends on it
@@ -35,8 +35,13 @@ Verifies #413
 
 flag_test
 ---------
-Test that a go_test_ rule that adds flags, even in the main package, can read
+Test that a `go_test`_ rule that adds flags, even in the main package, can read
 the flag.
 This does not even build a library, it's a test in the main package with no
 dependancies that checks it can declare and then read a flag.
 Verifies #838
+
+only_testmain_test
+------------------
+Test that an `go_test`_ that contains a ``TestMain`` function but no tests
+still builds and passes.

--- a/tests/core/go_test/only_testmain_test.go
+++ b/tests/core/go_test/only_testmain_test.go
@@ -1,0 +1,10 @@
+package only_testmain
+
+import (
+	"os"
+	"testing"
+)
+
+func TestMain(m *testing.M) {
+	os.Exit(m.Run())
+}

